// simple-react-calendar/lib/calendar/day.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _format = _interopRequireDefault(require("date-fns/format"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Day = function Day(props) {
  var blockClassName = props.blockClassName,
      date = props.date,
      handleOnClick = props.handleOnClick,
      handleOnEnter = props.handleOnEnter,
      isCurrentMonth = props.isCurrentMonth,
      isDisabled = props.isDisabled,
      isHighlighted = props.isHighlighted,
      isMonthNext = props.isMonthNext,
      isMonthPrev = props.isMonthPrev,
      isNonSelectable = props.isNonSelectable,
      isSelectable = props.isSelectable,
      isSelected = props.isSelected,
      isSelectionEnd = props.isSelectionEnd,
      isSelectionStart = props.isSelectionStart,
      isToday = props.isToday,
      isWeekend = props.isWeekend,
      isWorkday = props.isWorkday,
      customRender = props.customRender;
  var children = (0, _format.default)(date, 'D');

  if (customRender) {
    return customRender(_objectSpread({}, props, {
      children: children
    }));
  }

  return _react.default.createElement("button", {
    className: (0, _classnames.default)("".concat(blockClassName, "-day"), {
      'is-current_month': isCurrentMonth,
      'is-disabled': isDisabled,
      'is-end_selection': isSelectionEnd,
      'is-highlighted': isHighlighted,
      'is-next_month': isMonthNext,
      'is-not_selectable': isNonSelectable,
      'is-prev_month': isMonthPrev,
      'is-selectable': isSelectable,
      'is-selected': isSelected,
      'is-start_selection': isSelectionStart,
      'is-today': isToday,
      'is-weekend': isWeekend,
      'is-working_day': isWorkday
    }),
    onClick: handleOnClick,
    onMouseEnter: handleOnEnter,
    value: date
  }, children);
};

Day.propTypes = {
  blockClassName: _propTypes.default.string.isRequired,
  customRender: _propTypes.default.func,
  date: _propTypes.default.string.isRequired,
  handleOnClick: _propTypes.default.func,
  handleOnEnter: _propTypes.default.func,
  isCurrentMonth: _propTypes.default.bool.isRequired,
  isDisabled: _propTypes.default.bool.isRequired,
  isHighlighted: _propTypes.default.bool.isRequired,
  isMonthNext: _propTypes.default.bool.isRequired,
  isMonthPrev: _propTypes.default.bool.isRequired,
  isNonSelectable: _propTypes.default.bool.isRequired,
  isSelectable: _propTypes.default.bool.isRequired,
  isSelected: _propTypes.default.bool.isRequired,
  isSelectionEnd: _propTypes.default.bool.isRequired,
  isSelectionStart: _propTypes.default.bool.isRequired,
  isToday: _propTypes.default.bool.isRequired,
  isWeekend: _propTypes.default.bool.isRequired,
  isWorkday: _propTypes.default.bool.isRequired
};
var _default = Day;
exports.default = _default;
