// simple-react-calendar/lib/calendar/month.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _add_days = _interopRequireDefault(require("date-fns/add_days"));

var _are_ranges_overlapping = _interopRequireDefault(require("date-fns/are_ranges_overlapping"));

var _difference_in_calendar_days = _interopRequireDefault(require("date-fns/difference_in_calendar_days"));

var _end_of_month = _interopRequireDefault(require("date-fns/end_of_month"));

var _end_of_week = _interopRequireDefault(require("date-fns/end_of_week"));

var _is_before = _interopRequireDefault(require("date-fns/is_before"));

var _is_equal = _interopRequireDefault(require("date-fns/is_equal"));

var _is_same_day = _interopRequireDefault(require("date-fns/is_same_day"));

var _parse = _interopRequireDefault(require("date-fns/parse"));

var _start_of_month = _interopRequireDefault(require("date-fns/start_of_month"));

var _start_of_week = _interopRequireDefault(require("date-fns/start_of_week"));

var _sub_days = _interopRequireDefault(require("date-fns/sub_days"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _lib = require("./_lib");

var _consts = require("./consts");

var _days_of_week = _interopRequireDefault(require("./days_of_week"));

var _week = _interopRequireDefault(require("./week"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var RANGE_MODE = 'range';

var Month =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Month, _React$Component);

  function Month() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Month);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Month)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "handleOnDayMouseEnter", function (event) {
      event.preventDefault();
      var value = event.currentTarget.value;
      var date = (0, _parse.default)(value);
      var onDayMouseEnter = _this.props.onDayMouseEnter;

      if (onDayMouseEnter) {
        onDayMouseEnter(date);
      }

      if (!_this._selectionInProgress) return;
      var rangeLimit = _this.props.rangeLimit;
      var dateLimit = (0, _sub_days.default)(_this._selectionStart, rangeLimit);

      var isDisabledWithin = _this._getDisabledRange({
        start: (0, _is_before.default)(_this._selectionStart, date) ? _this._selectionStart : date,
        end: !(0, _is_before.default)(_this._selectionStart, date) ? _this._selectionStart : date
      });

      if (!isDisabledWithin) return;

      if (!(0, _is_equal.default)(date, _this._selectionEnd)) {
        if (!rangeLimit || rangeLimit && !(0, _is_before.default)(date, dateLimit)) {
          _this._selectionEnd = date;

          _this._pushUpdate();
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleOnDayClick", function (event) {
      event.preventDefault();
      var value = event.currentTarget.value;
      var date = (0, _parse.default)(value);
      var mode = _this.props.mode;

      if (mode === RANGE_MODE) {
        if (_this._selectionInProgress) {
          var isDisabledWithin = _this._getDisabledRange({
            start: (0, _is_before.default)(_this._selectionStart, date) ? _this._selectionStart : date,
            end: !(0, _is_before.default)(_this._selectionStart, date) ? _this._selectionStart : date
          });

          if (!isDisabledWithin) {
            _this._selectionInProgress = false;
            _this._selectionStart = null;
            _this._selectionEnd = null;

            _this._pushUpdate();

            _this._pushNoticeUpdate('overlapping_with_disabled');

            return;
          }

          _this._selectionInProgress = false;
          _this._selectionEnd = date;
        } else {
          _this._selectionInProgress = true;
          _this._selectionStart = date;
          _this._selectionEnd = date;
        }
      } else {
        _this._selectionInProgress = false;
        _this._selectionStart = date;
        _this._selectionEnd = date;
      }

      _this._pushUpdate();

      _this._pushNoticeUpdate(null);
    });

    _defineProperty(_assertThisInitialized(_this), "handleOnDisabledDayClick", function (event) {
      event.preventDefault();

      _this.props.onNoticeChange('disabled_day_click');
    });

    return _this;
  }

  _createClass(Month, [{
    key: "_pushUpdate",
    value: function _pushUpdate() {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          rangeLimit = _this$props.rangeLimit;
      var start, end;

      if (this._selectionStart && this._selectionEnd) {
        if ((0, _is_before.default)(this._selectionStart, this._selectionEnd)) {
          start = this._selectionStart;
          end = this._selectionEnd;
        } else {
          start = this._selectionEnd;
          end = this._selectionStart;
        }

        if (rangeLimit && rangeLimit < (0, _difference_in_calendar_days.default)(end, start)) {
          end = (0, _add_days.default)(start, rangeLimit);
        }
      }

      return onChange({
        start: start,
        end: end,
        inProgress: this._selectionInProgress
      });
    }
  }, {
    key: "_pushNoticeUpdate",
    value: function _pushNoticeUpdate(noticeType) {
      var onNoticeChange = this.props.onNoticeChange;
      return onNoticeChange(noticeType);
    }
  }, {
    key: "_getDisabledRange",
    value: function _getDisabledRange(interval) {
      var start = interval.start,
          end = interval.end;
      var disabledIntervals = this.props.disabledIntervals;
      if (!disabledIntervals) return true;

      for (var i = 0; i < disabledIntervals.length; i++) {
        var _disabledIntervals$i = disabledIntervals[i],
            intervalStart = _disabledIntervals$i.start,
            intervalEnd = _disabledIntervals$i.end;

        if ((0, _are_ranges_overlapping.default)(start, end, intervalStart, intervalEnd)) {
          return;
        }
      }

      return true;
    }
  }, {
    key: "_getMinDate",
    value: function _getMinDate() {
      var _this$props2 = this.props,
          rangeLimit = _this$props2.rangeLimit,
          minDate = _this$props2.minDate;
      var calcStartDate = (0, _sub_days.default)(this._selectionStart, rangeLimit);

      if (minDate) {
        var isCalcStartDayAfter = (0, _is_before.default)(minDate, calcStartDate);
        return isCalcStartDayAfter ? calcStartDate : minDate;
      } else {
        return calcStartDate;
      }
    }
  }, {
    key: "_getMaxDate",
    value: function _getMaxDate() {
      var _this$props3 = this.props,
          rangeLimit = _this$props3.rangeLimit,
          maxDate = _this$props3.maxDate;
      var calcEndDate = (0, _add_days.default)(this._selectionStart, rangeLimit);

      if (maxDate) {
        var isCalcEndDayBefore = (0, _is_before.default)(calcEndDate, maxDate);
        return isCalcEndDayBefore ? calcEndDate : maxDate;
      } else {
        return calcEndDate;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          blockClassName = _this$props4.blockClassName,
          customRender = _this$props4.customRender;

      var children = _react.default.createElement(_react.default.Fragment, null, this._renderDaysOfWeek(), this._renderWeeks());

      if (customRender) {
        return customRender(_objectSpread({}, this.props, {
          children: children
        }));
      }

      return _react.default.createElement("div", {
        className: "".concat(blockClassName, "-month")
      }, children);
    }
  }, {
    key: "_renderDaysOfWeek",
    value: function _renderDaysOfWeek() {
      var _this$props5 = this.props,
          disableDaysOfWeek = _this$props5.disableDaysOfWeek,
          blockClassName = _this$props5.blockClassName,
          weekStartsOn = _this$props5.weekStartsOn,
          daysOfWeek = _this$props5.daysOfWeek,
          renderDaysOfWeek = _this$props5.renderDaysOfWeek,
          renderDayOfWeek = _this$props5.renderDayOfWeek;
      if (disableDaysOfWeek) return;
      return _react.default.createElement(_days_of_week.default, {
        blockClassName: blockClassName,
        weekStartsOn: weekStartsOn,
        daysOfWeek: daysOfWeek,
        customRender: renderDaysOfWeek,
        renderDayOfWeek: renderDayOfWeek
      });
    }
  }, {
    key: "_renderWeeks",
    value: function _renderWeeks() {
      var _this2 = this;

      var _this$props6 = this.props,
          selectedMin = _this$props6.selectedMin,
          selectedMax = _this$props6.selectedMax,
          highlightedStart = _this$props6.highlightedStart,
          highlightedEnd = _this$props6.highlightedEnd,
          disabledIntervals = _this$props6.disabledIntervals,
          activeMonth = _this$props6.activeMonth,
          today = _this$props6.today,
          blockClassName = _this$props6.blockClassName,
          minNumberOfWeeks = _this$props6.minNumberOfWeeks,
          rangeLimit = _this$props6.rangeLimit,
          weekStartsOn = _this$props6.weekStartsOn,
          renderDay = _this$props6.renderDay,
          renderWeek = _this$props6.renderWeek;
      var weeks = [];
      var _this$props7 = this.props,
          minDate = _this$props7.minDate,
          maxDate = _this$props7.maxDate;
      var date = (0, _start_of_week.default)((0, _start_of_month.default)(activeMonth), {
        weekStartsOn: weekStartsOn
      });
      var end = (0, _end_of_week.default)((0, _end_of_month.default)(activeMonth), {
        weekStartsOn: weekStartsOn
      });

      if (this._selectionInProgress && rangeLimit) {
        minDate = this._getMinDate();
        maxDate = this._getMaxDate();
      }

      while (typeof minNumberOfWeeks == 'number' && minNumberOfWeeks > weeks.length || (0, _is_before.default)(date, end) || (0, _is_same_day.default)(date, end)) {
        weeks.push(date);
        date = (0, _add_days.default)(date, 7);
      }

      return weeks.map(function (week) {
        return _react.default.createElement(_week.default, {
          customRender: renderWeek,
          renderDay: renderDay,
          activeMonth: activeMonth,
          blockClassName: blockClassName,
          date: week,
          disabledIntervals: disabledIntervals,
          highlightedEnd: highlightedEnd,
          highlightedStart: highlightedStart,
          key: week.getTime(),
          maxDate: maxDate,
          minDate: minDate,
          onDayClick: _this2.handleOnDayClick,
          onDayMouseEnter: _this2.handleOnDayMouseEnter,
          onDisabledDayClick: _this2.handleOnDisabledDayClick,
          selectedMax: selectedMax,
          selectedMin: selectedMin,
          today: today,
          weekStartsOn: weekStartsOn
        });
      });
    }
  }]);

  return Month;
}(_react.default.Component);

exports.default = Month;

_defineProperty(Month, "propTypes", {
  activeMonth: _lib.datePropType.isRequired,
  blockClassName: _propTypes.default.string.isRequired,
  customRender: _propTypes.default.func,
  daysOfWeek: _propTypes.default.arrayOf(_propTypes.default.string),
  disableDaysOfWeek: _propTypes.default.bool,
  disabledIntervals: _propTypes.default.arrayOf(_propTypes.default.shape({
    start: _lib.datePropType.isRequired,
    end: _lib.datePropType.isRequired
  })),
  highlightedEnd: _lib.datePropType,
  highlightedStart: _lib.datePropType,
  maxDate: _lib.datePropType,
  minDate: _lib.datePropType,
  minNumberOfWeeks: _propTypes.default.number,
  mode: _propTypes.default.string.isRequired,
  onChange: _propTypes.default.func.isRequired,
  onDayMouseEnter: _propTypes.default.func,
  onNoticeChange: _propTypes.default.func.isRequired,
  rangeLimit: _propTypes.default.number,
  renderDay: _propTypes.default.func,
  renderDayOfWeek: _propTypes.default.func,
  renderDaysOfWeek: _propTypes.default.func,
  renderWeek: _propTypes.default.func,
  selectedMax: _lib.datePropType,
  selectedMin: _lib.datePropType,
  today: _lib.datePropType.isRequired,
  weekStartsOn: _propTypes.default.oneOf(_consts.DAYS_IN_WEEK)
});
