// simple-react-calendar/lib/calendar/week.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _each_day = _interopRequireDefault(require("date-fns/each_day"));

var _end_of_week = _interopRequireDefault(require("date-fns/end_of_week"));

var _format = _interopRequireDefault(require("date-fns/format"));

var _is_after = _interopRequireDefault(require("date-fns/is_after"));

var _is_before = _interopRequireDefault(require("date-fns/is_before"));

var _is_equal = _interopRequireDefault(require("date-fns/is_equal"));

var _is_same_day = _interopRequireDefault(require("date-fns/is_same_day"));

var _is_same_month = _interopRequireDefault(require("date-fns/is_same_month"));

var _is_weekend = _interopRequireDefault(require("date-fns/is_weekend"));

var _is_within_range = _interopRequireDefault(require("date-fns/is_within_range"));

var _start_of_day = _interopRequireDefault(require("date-fns/start_of_day"));

var _start_of_week = _interopRequireDefault(require("date-fns/start_of_week"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _lib = require("./_lib");

var _consts = require("./consts");

var _day = _interopRequireDefault(require("./day"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Week =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Week, _React$Component);

  function Week() {
    _classCallCheck(this, Week);

    return _possibleConstructorReturn(this, _getPrototypeOf(Week).apply(this, arguments));
  }

  _createClass(Week, [{
    key: "_dateSelectable",
    value: function _dateSelectable(date) {
      var _this$props = this.props,
          minDate = _this$props.minDate,
          maxDate = _this$props.maxDate;

      if (this._dateDisabled(date)) {
        return false;
      }

      if (minDate && maxDate) {
        return (0, _is_within_range.default)(date, minDate, maxDate);
      } else if (minDate && !maxDate) {
        return (0, _is_after.default)(date, minDate) || (0, _is_equal.default)(date, minDate);
      } else if (maxDate && !minDate) {
        return (0, _is_before.default)(date, maxDate) || (0, _is_equal.default)(date, maxDate);
      } else {
        return true;
      }
    }
  }, {
    key: "_dateSelected",
    value: function _dateSelected(date) {
      var _this$props2 = this.props,
          selectedMin = _this$props2.selectedMin,
          selectedMax = _this$props2.selectedMax;
      return Boolean(selectedMin && selectedMax && (0, _is_within_range.default)((0, _start_of_day.default)(date), (0, _start_of_day.default)(selectedMin), (0, _start_of_day.default)(selectedMax)));
    }
  }, {
    key: "_dateHighlighted",
    value: function _dateHighlighted(date) {
      var _this$props3 = this.props,
          highlightedStart = _this$props3.highlightedStart,
          highlightedEnd = _this$props3.highlightedEnd;
      if (!highlightedStart || !highlightedEnd) return false;
      return (0, _is_within_range.default)((0, _start_of_day.default)(date), (0, _start_of_day.default)(highlightedStart), (0, _start_of_day.default)(highlightedEnd));
    }
  }, {
    key: "_dateDisabled",
    value: function _dateDisabled(date) {
      var dateDisabled;
      var disabledIntervals = this.props.disabledIntervals;
      if (!disabledIntervals) return false;

      for (var i = 0; i < disabledIntervals.length; i++) {
        var _disabledIntervals$i = disabledIntervals[i],
            start = _disabledIntervals$i.start,
            end = _disabledIntervals$i.end;
        dateDisabled = (0, _is_within_range.default)((0, _start_of_day.default)(date), (0, _start_of_day.default)(start), (0, _start_of_day.default)(end));

        if (dateDisabled) {
          return dateDisabled;
        }
      }

      return false;
    }
  }, {
    key: "render",
    value: function render() {
      var customRender = this.props.customRender;

      var children = this._renderDays();

      if (customRender) {
        return customRender(_objectSpread({}, this.props, {
          children: children
        }));
      }

      return _react.default.createElement("div", {
        className: "".concat(this.props.blockClassName, "-week")
      }, children);
    }
  }, {
    key: "_renderDays",
    value: function _renderDays() {
      var _this = this;

      var _this$props4 = this.props,
          date = _this$props4.date,
          today = _this$props4.today,
          onDayClick = _this$props4.onDayClick,
          onDisabledDayClick = _this$props4.onDisabledDayClick,
          onDayMouseEnter = _this$props4.onDayMouseEnter,
          blockClassName = _this$props4.blockClassName,
          activeMonth = _this$props4.activeMonth,
          selectedMax = _this$props4.selectedMax,
          selectedMin = _this$props4.selectedMin,
          weekStartsOn = _this$props4.weekStartsOn,
          renderDay = _this$props4.renderDay;
      var start = (0, _start_of_week.default)(date, {
        weekStartsOn: weekStartsOn
      });
      var end = (0, _end_of_week.default)(date, {
        weekStartsOn: weekStartsOn
      });
      return (0, _each_day.default)(start, end).map(function (day) {
        var date = (0, _format.default)(day, 'YYYY-MM-DD');

        var isSelectable = _this._dateSelectable(day);

        var isDisabled = _this._dateDisabled(date);

        var isWorkDay = !(0, _is_weekend.default)(date);
        var isCurrentMonth = (0, _is_same_month.default)(date, activeMonth);
        var isNextMonth = !isCurrentMonth && (0, _is_after.default)(date, activeMonth);
        return _react.default.createElement(_day.default, {
          customRender: renderDay,
          blockClassName: blockClassName,
          date: date,
          handleOnClick: isSelectable ? onDayClick : isDisabled ? onDisabledDayClick : null,
          handleOnEnter: isSelectable ? onDayMouseEnter : null,
          isCurrentMonth: isCurrentMonth,
          isDisabled: isDisabled,
          isHighlighted: _this._dateHighlighted(day),
          isMonthNext: isNextMonth,
          isMonthPrev: !isCurrentMonth && !isNextMonth,
          isNonSelectable: !isSelectable,
          isSelectable: isSelectable,
          isSelected: _this._dateSelected(day),
          isSelectionEnd: Boolean(selectedMax && (0, _is_same_day.default)(selectedMax, day)),
          isSelectionStart: Boolean(selectedMin && (0, _is_same_day.default)(selectedMin, day)),
          isToday: (0, _is_same_day.default)(today, day),
          isWeekend: !isWorkDay,
          isWorkday: isWorkDay,
          key: date
        });
      });
    }
  }]);

  return Week;
}(_react.default.Component);

exports.default = Week;

_defineProperty(Week, "propTypes", {
  activeMonth: _lib.datePropType.isRequired,
  blockClassName: _propTypes.default.string.isRequired,
  customRender: _propTypes.default.func,
  date: _lib.datePropType.isRequired,
  disabledIntervals: _propTypes.default.arrayOf(_propTypes.default.shape({
    start: _lib.datePropType.isRequired,
    end: _lib.datePropType.isRequired
  })),
  highlightedEnd: _lib.datePropType,
  highlightedStart: _lib.datePropType,
  maxDate: _lib.datePropType,
  minDate: _lib.datePropType,
  onDayClick: _propTypes.default.func.isRequired,
  onDayMouseEnter: _propTypes.default.func.isRequired,
  onDisabledDayClick: _propTypes.default.func.isRequired,
  renderDay: _propTypes.default.func,
  selectedMax: _lib.datePropType,
  selectedMin: _lib.datePropType,
  today: _lib.datePropType.isRequired,
  weekStartsOn: _propTypes.default.oneOf(_consts.DAYS_IN_WEEK)
});
